overhangio/tutor
# feat: k8s horizontal pod autoscaling

## Description

This PR adds capability for autoscaling for lms, studio and their workers.The autoscaling configuration is disabled by default to not make it backward compatible and must be enabled one-by-one.

Also, the PR refactors resource request to make it configurable for lms and studio. The default value is the same as the hardcoded value was.

Since the default configuration values are sorted (again), this make it harder to see what's the addition. The following **new** config options were added:

**Autoscaling for studio**

* `CMS_AUTOSCALING`
* `CMS_AVG_CPU`
* `CMS_AVG_MEMORY`
* `CMS_CPU_LIMIT`
* `CMS_MAX_REPLICAS`
* `CMS_MEMORY_LIMIT`
* `CMS_MIN_REPLICAS`

**Autoscaling for studio workers**

* `CMS_WORKER_AUTOSCALING`
* `CMS_WORKER_AVG_CPU`
* `CMS_WORKER_AVG_MEMORY`
* `CMS_WORKER_CPU_LIMIT`
* `CMS_WORKER_MAX_REPLICAS`
* `CMS_WORKER_MEMORY_LIMIT`
* `CMS_WORKER_MIN_REPLICAS`

**Autoscaling for lms**

* `LMS_AUTOSCALING`
* `LMS_AVG_CPU`
* `LMS_AVG_MEMORY`
* `LMS_CPU_LIMIT`
* `LMS_MAX_REPLICAS`
* `LMS_MEMORY_LIMIT`
* `LMS_MIN_REPLICAS`

**Autoscaling for lms workers**

* `LMS_WORKER_AUTOSCALING`
* `LMS_WORKER_AVG_CPU`
* `LMS_WORKER_AVG_MEMORY`
* `LMS_WORKER_CPU_LIMIT`
* `LMS_WORKER_MAX_REPLICAS`
* `LMS_WORKER_MEMORY_LIMIT`
* `LMS_WORKER_MIN_REPLICAS`

**Resource request config**

* `CMS_MEMORY_REQUEST`
* `LMS_MEMORY_REQUEST`

## Supporting information

At OpenCraft, we already tried it. Here are the results:

<img width="891" alt="Screenshot 2022-05-30 at 8 15 47" src="https://user-images.githubusercontent.com/19173947/170928730-97b9f734-6c75-4f50-a5eb-14cc469b7982.png">


### Dependencies

N/A

## Testing instructions

* Proofread the changes
* Setup a new instance with autoscaling enabled
* Run `kubectl get hpa` in the instance's namespace
* Validate the HPA list contains the expected results

## Deadline

None

## Other information

N/A
